src/main/java/org/allenai/wikitables/WikiTablesExample.java

```diff
@@ -4,41 +4,85 @@
-    // We are making the logicalForm in CcgExample null, and defining the set logicalForms in this class.
-    super(sentence, null, null, null);
+  private final AnnotatedSentence sentence;
+  private final Set<Expression2> logicalForms;
+  private final ContextValue context;
+  private final Value targetValue;
+
+  public WikiTablesExample(AnnotatedSentence sentence, Set<Expression2> logicalForms, ContextValue context,
+                           Value targetValue) {
+    this.sentence = sentence;
     this.logicalForms = logicalForms;
+    this.context = context;
+    this.targetValue = targetValue;
   }
 
   public Set<Expression2> getLogicalForms() {
     return logicalForms;
   }
 
-  @Override
-  public boolean hasLogicalForm() {
-    return logicalForms != null;
+  public AnnotatedSentence getSentence() {
+    return sentence;
+  }
+
+  public boolean isFormulaCorrect(Expression2 pnpFormula) {
+    // Sempre represents lambda expressions differently. We changed them when reading the examples. Changing
+    // them back for execution.
+    System.out.println("Before conversion: " + pnpFormula.toString());
+    String expressionString = toSempreLambdaForm(pnpFormula.toString());
+    System.out.println("After conversion: " + expressionString);
+    try {
+      Formula sempreFormula = Formula.fromString(expressionString);
+      return WikiTablesDataProcessor.isFormulaCorrect(sempreFormula, context, targetValue, null);
+    } catch (Exception ex) {
+      System.err.println("Bad formula: " + expressionString);
+      return false;
+    }
   }
 
-  @Override
   public String toString() {
-    String superString = super.toString();
     List<String> logicalFormStrings = new ArrayList<>();
     for (Expression2 lf : logicalForms) {
         logicalFormStrings.add(lf.toString());
     }
-    return superString + " [\n" + String.join(" ", logicalFormStrings) + "\n]";
+    return sentence + " [\n" + String.join(" ", logicalFormStrings) + "\n]";
+  }
+
+  public static String toPnpLambdaForm(String expression) {
+    /*
+    Sempre's lambda expressions are written differently from what pnp expects. We make the following change
+    (lambda x ((reverse fb:cell.cell.number) (var x))) -> (lambda (x) ((reverse fb:cell.cell.number) x))
+     */
+    // Sempre's lambda expressions have only one free variable, I think.
+    expression = expression.replaceAll("lambda x", "lambda (x)");
+    expression = expression.replaceAll("(var x)", "x");
+    return expression;
+  }
+
+  public static String toSempreLambdaForm(String expression) {
+    // TODO: Properly de-canonicalize variable names, to work with multiple variables.
+    expression = expression.replaceAll("lambda \\(\\$0\\)", "lambda x");
+    // Remove single sub-expressions within parens. Eg: (fb:type.object.type (fb:type.row)) -> (fb:type.object.type fb:type.row)
+    expression = expression.replaceAll("\\(([^ ]*)\\)", "$1");
```

Comment:
This conversion is incorrect. (fb:type.row) is not equivalent to fb:type.row. One of these presumably shouldn't be generated by the parser, and the problem should be fixed there and not here.